lib/core/facets/array/subtract.rb

```diff
@@ -0,0 +1,37 @@
+class Array
+  # Returns a new array that is a copy of the original array, removing the _first_ item in self that
+  # also appears in other. The order is preserved from the original array.
+  #
+  # For each element of `other`, deletes the _first_ item from self that is equal to element (using
+  # `delete_at`, not `delete`).
+  #
+  # This is similar to `Array#-` and `Array#difference`, except that instead of removing _all_
+  # matches, it only removes as many occurrences as you actually _ask_ it to subtract:
+  #
+  #    > [1, 1, 2].subtract [1]
+  #   => [1, 2]
+  #
+  #    > [1, 1, 2].subtract [1, 1]
+  #   => [2]
+  #
+  #    > [1, 1, 2] - [1]
+  #   => [2]
+  #
+  #    > [1, 1, 2].difference [1]
+  #   => [2]
+  #
+  def subtract(other)
+    dup.subtract!(other)
+  end
+
+  # Removes the _first_ item in self that also appears in other.
+  #
+  # This is the in-place version of subtract.
+  #
+  def subtract!(other)
+    other.each do |el|
+      delete_at(index(el))
```

Comment:
This will be erroed in this case:

```ruby
[1, 2, 3].substract! [4]

TypeError (no implicit conversion from nil to integer)
```
Since you are looking for the element A[i] in B, `index` returns `nil` when not found, and `delete_at` will complain

Reply:
Good catch! I've pushed up a fix (extracted new `delete_first` method)...